elasticai/creator/nn/fixed_point/linear/testbench.py

```diff
@@ -0,0 +1,122 @@
+from typing import Protocol
+from collections import defaultdict
+
+from elasticai.creator.vhdl.design.ports import Port
+from elasticai.creator.file_generation.savable import Path
+from elasticai.creator.file_generation.template import (
+    InProjectTemplate,
+    module_to_package,
+)
+
+from elasticai.creator.nn.fixed_point.number_converter import NumberConverter, FXPParams
+import math
+
+
+class LinearDesignProtocol(Protocol):
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        ...
+
+    @property
+    @abstractmethod
+    def port(self) -> Port:
+        ...
+
+    @property
+    @abstractmethod
+    def in_feature_num(self) -> int:
+        ...
+    @property
+    @abstractmethod
+    def out_feature_num(self) -> int:
+        ...
+
+    @property
+    @abstractmethod
+    def frac_width(self) -> int:
+        ...
+
+    @property
+    @abstractmethod
+    def data_width(self) -> int:
+        ...
+
+
+class LinearTestbench:
+    def __init__(self, name: str, uut: LinearDesignProtocol):
+        self._converter_for_batch = NumberConverter(FXPParams(8, 0))  # max for 255 lines of inputs
+        self._name = name
+        self._uut_name = uut.name
+        self._input_signal_length = uut.in_feature_num
+        self._x_address_width = uut.port["x_address"].width
+        self._fxp_params = FXPParams(uut.data_width, uut.frac_width)
+        self._converter = NumberConverter(self._fxp_params)
+        self._output_signal_length = uut.out_feature_num
+        self._y_address_width = uut.port["y_address"].width
+
+    def save_to(self, destination: Path):
+        template = InProjectTemplate(
+            package=module_to_package(self.__module__),
+            file_name="testbench.tpl.vhd",
+            parameters={
+                "testbench_name": self.name,
+                "input_signal_length": str(self._input_signal_length),
+                "total_bits": str(self._fxp_params.total_bits),
+                "x_address_width": str(self._x_address_width),
+                "output_signal_length": str(self._output_signal_length),
+                "y_address_width": str(self._y_address_width),
+                "uut_name": self._uut_name,
+            },
+        )
+        destination.create_subpath(self.name).as_file(".vhd").write(template)
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    def prepare_inputs(self, *inputs) -> list[dict]:
+        batches = inputs[0]
+        prepared_inputs = []
+        for batch in batches:
+            prepared_inputs.append({})
+            for channel_id, channel in enumerate(batch):
+                for time_step_id, time_step_val in enumerate(channel):
+                    prepared_inputs[-1][f"x_{channel_id}_{time_step_id}"] = (
+                        self._converter.rational_to_bits(time_step_val)
+                    )
+        return prepared_inputs
+
+    def parse_reported_content(self, content: list[str]) -> list[list[list[float]]]:
```

Comment:
This function is hard to understand, it is even harder to deduct the output format that is expected from the test bench. Thus, i'd like to see a comment that outlines the expected format. That should also help refactoring the function (or adding tests)

Reply:
I added comments for the functions.